File: C++/6. ZigZag Conversion.cpp
class Solution {
public:
    string convert(string s, int numRows) {
        if (numRows == 1 || s.size() < 2)
            return s;
        int interval = numRows - 2;
        string res;
        //find a relationship in (numRows,interval,i,s_pos)
        for (int i = 0; i < numRows; i++) { // row by row
            for (int s_pos = i; s_pos < s.size(); ) {
                res += s[s_pos];//the position that every Column was full ,e.p. PAHN
                s_pos = s_pos + numRows + interval;//the next position that every Column was full
                if (i != 0 && i != numRows - 1) { // deal with zigzap shape
                    if (s_pos - (2 * i) < s.size())
                        res += s[s_pos - (2 * i)]; // the position only 1 character in column
                }
            }
        }

        return res; 
    }
};

File: Python/6. ZigZag Conversion.py
class Solution:
    def convert(self, s, numRows):
        """
        :type s: str
        :type numRows: int
        :rtype: str
        """
        s_length = len(s)
        if numRows == 1 or s_length < 2:  # numrows too small or s too short
            return s

        interval = numRows - 2
        res = ""
        for i in range(numRows):  # for each row
            s_pos = i
            while s_pos < s_length:
                res += s[s_pos]  # add full column character postion
                s_pos = s_pos + numRows + interval  # next full column character

                if i != 0 and i != numRows-1:  # in zigzag shape character
                    if s_pos - (2*i) < s_length:
                        res += s[s_pos - (2*i)]  # add zigzag shape character

        return res
